Make set_order_props_from_data() protected rather than private

james-allan opened the pull request:
### All Submissions:

-   [X] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [X] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This changes the `set_order_props_from_data()` function to be protected rather than private. 

This fairly simple change, enables us in WC Subscriptions to override this function so we can set our own subscription specific properties while the subscription object is being read. This fixes issues where loading a subscription was incorrectly causing it to resync because WC would determine that the HPOS loaded object was different from the post record because the post object is completely loaded but the HPOS subscription object hasn't finished loading because we need to set subscription specific properties and that was only possible to run after `read_multiple()` and so after the WC comparison was done.

Because all the `read_multiple()` functions are private, there is no opportunity for us to set our own subscription properties without reimplementing large parts of the order datastore. 

This PR enables us to:

1. Override the `set_order_props_from_data()` function in the Subscription datastore. 
2. Inside that function populate our custom properties from the given data fetched from the database. 

All while the subscription is being read and before WC compares the order/subscription to see if it needs a resync. 

I've submitted the required changes to WC Subscriptions here: https://github.com/Automattic/woocommerce-subscriptions-core/pull/336

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [x] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

james-allan committed:
Make set_order_props_from_data() protected

This enables 3rd parties that extend the order datastore and need to set their own data. For example, Subscriptions

james-allan committed:
Add changelog entry

james-allan commented:
I should add that if you're opposed to making this function `protected`, we're open to other suggestions that would enable us to set our subscription properties while it is being read and before WC core determines if a resync needs to occur. Eg a filter or something.

I personally prefer this approach because adding a filter opens us up to much broader use (anyone can hook into it), and a filter would require us to use a weird code pattern to use a filter to set core properties, when the datastores don't use hooks in that way. 

While the name of this function (`set_order_props_from_data()`) is a little strange in a subscriptions context, it made the most sense given this function matches with what we're doing too - setting core properties, ours just happen to come from the meta table still.

vedanshujain commented:
@james-allan , what do you think making `init_order_record` and `get_order_data_for_ids` protected instead of `set_order_props_from_data`? Will it solve your use case?

I am asking because `set_order_props_from_data` is really not a standalone method in itself, it depends on order data being passed, metadata to be already initialized and order_id to be already set. Whereas both `init_order_record` and `get_order_data_for_ids` are more standalone methods, can be replaced wholly (or extended), and therefore provide better extensibility.

james-allan commented:
@vedanshujain that would work for us. 

A couple of points:

1. In our case we wouldn't need to use the `get_order_data_for_ids()` function if it was protected. That function already fetches the data we need (under the meta data arg), but I take your point about these being linked to provide better extensibility for third-parties to: 
    a. override/extend what data is fetched and from where,
    b. and then override/extend the corresponding function that handles that data.
3. The `init_order_record()` function flags the object as read at the bottom. This isn't ideal of us but we can work with it. To explain with an example, we would need to do the following:

```php
protected function init_order_record( $subscription, $subscription_id, $subscription_data ) {
    parent::init_order_record( $subscription, $subscription_id, $subscription_data );
    $subscription->set_object_read( false );

    // Do all our setting of subscription specific data

   $subscription->set_object_read( true );
}
```

james-allan commented:
I've submitted the changes in [f40cfd8](https://github.com/james-allan/woocommerce/commit/f40cfd887c832e7bfc1a85b01b3d757b4e46819f) and [337085e](https://github.com/james-allan/woocommerce/commit/337085e11242ad1506c0fa34573d94c1dbb90142), however they are appearing on a fork because I lost write access to this Repo. 

I'll see if one of my colleagues is able to submit those changes on my behalf.

Commit:
return set_order_props_from_data to private visibility

Commit:
Make init_order_record and get_order_data_for_ids protected

Following feedback here: https://github.com/woocommerce/woocommerce/pull/35829#issuecomment-1340528244

Commit:
Update changelog entry

vedanshujain reviewed (approved):
LGTM! Thanks!
